### PR TITLE
Un-deprecate KeepAliveHandle.release()

### LINKED ISSUE #123712: `KeepAliveHandle.release()` should not be deprecated, or a better-named alternative should be introduced
## Use case
In 3.3.0, `KeepAliveHandle.release()` was deprecated (see #108384).
However, in some situations, it makes sense to just create a single `KeepAliveHandle` for the entire lifecycle of an `Element`, and then just `dispose()` it once when the applicable `Element` is disposed. Then, in `deactivate`, simply call `release` on the `KeepAliveHandle`. This is the intended behavior as documented at https://api.flutter.dev/flutter/widgets/KeepAliveNotification-class.html and https://api.flutter.dev/flutter/widgets/KeepAliveHandle-class.html.

## Proposal
Seeing as the `KeepAliveHandle` docs indicate it can be manually triggered without a need for disposal (which makes sense), this behavior should be preserved and not deprecated.
The creation of a new method equivalent to release (but perhaps with a more declarative name) or removing the deprecation status on `release` would solve this. I'd be happy to open a PR for this.

Suggested new method name: `removeKeepAlive`. This name does not imply it is disposing nor releasing any resources. In addition, documentation can be added to this new method that indicates `dispose` *must be called* later on.
```dart
class KeepAliveHandle extends ChangeNotifier {
  // ...

  /// Trigger the listeners to indicate that the widget
  /// no longer needs to be kept alive. A widget can be marked as needing to be
  /// kept alive again with [KeepAliveNotification.dispatch].
  ///
  /// *This method does not call [dispose].* When the handle is not needed
  /// anymore, it must be [dispose]d regardless of whether notifying listeners.
  void removeKeepAlive() {
    notifyListeners();
  }

  // ...
}
```

### PR DESCRIPTION
**Pull request by GregoryConrad**
In 3.3.0, `KeepAliveHandle.release()` was deprecated (see #108384).
However, in many situations, it makes sense to just create a single `KeepAliveHandle` for the entire lifecycle of an `Element`, and then just `dispose()` it once when the applicable `Element` is disposed. Then, in `deactivate`, simply call `release` (or an equivalently named method) on the `KeepAliveHandle`. This is the intended behavior as documented at https://api.flutter.dev/flutter/widgets/KeepAliveNotification-class.html and https://api.flutter.dev/flutter/widgets/KeepAliveHandle-class.html.

Fixes #123712 by adding in a `removeKeepAlive` method that is more clear than `release`.

### CODE REVIEW

**Comment by GregoryConrad** [comment]
CC @dnfield for your work in #108384 (if you are not an appropriate person to review this, please add the right person!)

**Comment by GregoryConrad** [comment]
Just realized the current implementation (on `master` right now) has a memory leak if a `KeepAliveHandle` is used in a new part of the Widget tree/outlives the ancestor `AutomaticKeepAlive` (it'll hold onto a reference of the old `AutomaticKeepAlive`). Fixing that as well in this PR.

**Comment by dnfield** [comment]
The main concern is that developers do not call `removeKeepAlive` and fail to call `dispose`, as that will result in unexpectedly retaining memory.

Could you provide an example of an element that wants to reuse this between activate/deactivate/dispose lifecycles?

**Comment by GregoryConrad** [comment]
> The main concern is that developers do not call removeKeepAlive and fail to call dispose, as that will result in unexpectedly retaining memory.

@dnfield Agreed, I figured that the deprecation of `release` would help with that so that anyone using `release` incorrectly at the moment will now know the correct usage. The name `removeKeepAlive` also doesn't imply a `dispose` like `release` might (i.e., *release* a resource), which is why I chose that name. I also made some other points as to why this is helpful over in #123712. But, I think through documentation, it should be clear that `dispose` must also be called.

> Could you provide an example of an element that wants to reuse this between activate/deactivate/dispose lifecycles?

I covered a few cases/reasons here: https://github.com/flutter/flutter/issues/123712#issuecomment-1490419430

**Comment by dnfield** [comment]
If `AutomaticKeepAliveMixin` benefits from this, it should get updated with this change. I think it would help to see what that change looks like. My main concern is over whether it's going to be easier to incorrectly use the object now, or forget to dispose it. Forgetting to dispose it can have serious consequences on applications.

**Comment by GregoryConrad** [comment]
Added in the mixin changes as requested, but GitHub's online UI really just messed up a rebase. Will try to fix that in a second.

File: packages/flutter/lib/src/widgets/automatic_keep_alive.dart
Comment (christopherfujino) [inline packages/flutter/lib/src/widgets/automatic_keep_alive.dart:147]:
```suggestion
    final VoidCallback callback;
```

Comment (GregoryConrad) [reply packages/flutter/lib/src/widgets/automatic_keep_alive.dart:149]: I needed to add the `late` because Dart doesn't know for sure that `callback` will be initialized when it is used within itself (`handle.removeListener(callback);`):

> The final variable 'callback' can't be read because it's potentially unassigned at this point.

Comment (christopherfujino) [reply packages/flutter/lib/src/widgets/automatic_keep_alive.dart:149]: ahh, I see

**Review by Piinks — COMMENTED** [review]
Hey @GregoryConrad welcome, and thanks for contributing!

It looks like from the commits and the comments, this PR is trying to fix multiple things. We typically split those out into individual PRs rather than lumping them all together.
This helps us make sure each individual change is well tested. It also ensures if one of the changes causes an issue and is reverted, we don't have to revert all of the changes as a result of one of them causing a problem.

Specific to the deprecated method, if deprecating the method was not a beneficial change, we should consider un-deprecating it instead of adding it back under a new name. This may be clearer when the code is separated from other changes, but that is my initial impression. :)

**Comment by GregoryConrad** [comment]
@Piinks thanks for the feedback! Originally, this PR was just meant to add a new method, `removeKeepAlive`, instead of `release`, which was deprecated. Reason being that `release` is not all too descriptive, and a user might think `release` actually cleans up its resources as well (i.e. *releases them*, which it does not). Thus, the name `release` itself is a bit of an issue, which is why I introduced `removeKeepAlive` instead of un-deprecating `release`.

The PR got a bit larger because there were a handful of issues in the current implementation on `master` (inconsistent with the documentation) that I needed to fix in order for `removeKeepAlive` to work. Per your suggestion, I'll toss those in a new PR!

**Comment by Piinks** [comment]
Hey @GregoryConrad did you want to update this PR and remove the changes that were separated out to other PRs?
If the main purpose here is to replace the deprecated method with the same one of a different name, we should just un-deprecate the original method. Renaming something is not typically a breaking change we 
 subject developers to.

**Comment by GregoryConrad** [comment]
@Piinks Sure, I'll update this PR and un-deprecate the original method. Please note that CI here will fail until #124168 is merged.

**Comment by Piinks** [comment]
> Please note that CI here will fail until https://github.com/flutter/flutter/pull/124168 is merged.

Why is that? Is this not just a revert of https://github.com/flutter/flutter/pull/108384?

**Comment by GregoryConrad** [comment]
@Piinks It's not just a revert; per https://github.com/flutter/flutter/pull/123864#issuecomment-1496222327, I changed the `AutomaticKeepAliveMixin` to use `release` alongside `dispose`. It makes the implementation a lot easier to follow (if you look at `AutomaticKeepAliveMixin` before/after, you will see what I am talking about--the diffs shown on GitHub don't quite do it justice). However, using `release` alongside `dispose` without keeping track of an extra mutable variable within the mixin requires #124168 in order to function correctly.

In the end, if the team decides to not merge this PR due to the confusing naming of `release` vs `dispose` as @dnfield pointed out previously, that makes sense with me and I'd be ok with that since I can just get make my own `Listenable` implementation in my own project to use with `KeepAliveNotification` (it's nice to have that flexibility). However, it would still be nice to have #124168 because the current API is overly restrictive for (as far as I can tell) no good reason. It is currently impossible to use keep alives without needing to maintain mutability somewhere, unless #124168 is merged.

**Comment by Piinks** [comment]
Dan is OOO this week, so I would like to wait and get his inout on this.
This is not just un-deprecating the method, it results in different functionality from before the deprecation was added and I find that concerning from a developer perspective that won't otherwise understand that this has changed.

**Comment by GregoryConrad** [comment]
> it results in different functionality from before the deprecation was added

Sorry; wasn’t clear enough here. These changes are completely invisible to any developers using keep alives. In addition to un-deprecating release, this PR has a slight refactor of the KeepAliveMixin to actually use release and to be more maintainable/readable. #124168 also maintains 100% backwards compatibility, but eliminates a situation which would previously throw an unintuitive assert.

**Review by dnfield — CHANGES_REQUESTED** [review]
I have the same concerns with this as what I mentioned in https://github.com/flutter/flutter/pull/124168#pullrequestreview-1398870193

Allowing this encourages behavior that the framework currently documents as bad behavior. Reusing objects that need to be disposed is a strange pattern that we don't really use elsewhere AFAICT. 

I think this needs some more discussion if we'r egoing to make it work, but I really don't see how this is landable in its current state.

**Comment by dnfield** [comment]
Closing this to remove it from review queues - again, please feel free to reach out on discord if you want to discuss further.

**Pull request closed without merging**
